From 7f338ed9308ee0364dd371ad022506f42ff8c89d Mon Sep 17 00:00:00 2001
From: "Huang, Xiaojun" <xiaojun.huang@intel.com>
Date: Mon, 13 Nov 2023 21:08:25 +0800
Subject: [PATCH] ipcl patch

Signed-off-by: Huang, Xiaojun <xiaojun.huang@intel.com>
---
 CMakeLists.txt      | 21 +++++++++++++++++----
 ipcl/CMakeLists.txt | 32 ++++++++++++++++----------------
 2 files changed, 33 insertions(+), 20 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index db92edf..88e3fe3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -194,8 +194,8 @@ set(IPCL_FORWARD_CMAKE_ARGS
 set(IPCL_ROOT_DIR ${CMAKE_CURRENT_LIST_DIR})
 set(IPCL_SRC_DIR ${IPCL_ROOT_DIR}/ipcl)
 set(IPCL_INC_DIR ${IPCL_SRC_DIR}/include)
-set(IPCL_INSTALL_INCLUDEDIR ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_INCLUDEDIR}/ipcl)
-set(IPCL_INSTALL_LIBDIR ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}/ipcl)
+set(IPCL_INSTALL_INCLUDEDIR ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_INCLUDEDIR})
+set(IPCL_INSTALL_LIBDIR ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR})
 set(IPCL_CMAKE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake/ipcl")
 if(CMAKE_BUILD_TYPE STREQUAL "Debug")
   set(IPCL_DEBUG ON)
@@ -211,8 +211,21 @@ find_package(Threads REQUIRED)
 find_package(OpenSSL REQUIRED)
 
 # External dependencies
-include(cmake/ippcrypto.cmake)
-include(cmake/cereal.cmake)
+# include(cmake/ippcrypto.cmake)
+# include(cmake/cereal.cmake)
+
+add_library(IPPCP::ippcp STATIC IMPORTED GLOBAL)
+add_library(IPPCP::crypto_mb STATIC IMPORTED GLOBAL)
+
+set_target_properties(IPPCP::ippcp PROPERTIES
+          IMPORTED_LOCATION ${IPPCRYPTO_LIB_DIR}/libippcp.a
+          INCLUDE_DIRECTORIES ${IPPCRYPTO_INC_DIR}
+)
+
+set_target_properties(IPPCP::crypto_mb PROPERTIES
+          IMPORTED_LOCATION ${IPPCRYPTO_LIB_DIR}/libcrypto_mb.a
+          INCLUDE_DIRECTORIES ${IPPCRYPTO_INC_DIR}
+)
 
 if(IPCL_DETECT_CPU_RUNTIME)
   include(cmake/cpufeatures.cmake)
diff --git a/ipcl/CMakeLists.txt b/ipcl/CMakeLists.txt
index a338de1..c933631 100644
--- a/ipcl/CMakeLists.txt
+++ b/ipcl/CMakeLists.txt
@@ -43,35 +43,35 @@ install(DIRECTORY ${IPCL_INC_DIR}/
 )
 
 # CEREAL (third party dep): include and install definition
-add_dependencies(ipcl ext_cereal)
+# add_dependencies(ipcl ext_cereal)
 target_include_directories(ipcl
     PUBLIC $<BUILD_INTERFACE:${CEREAL_INC_DIR}>
 	PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/ipcl>
 )
 
-install(DIRECTORY ${CEREAL_INC_DIR}/
-    DESTINATION ${IPCL_INSTALL_INCLUDEDIR}
-    FILES_MATCHING
-    PATTERN "*.hpp"
-    PATTERN "*.h"
-)
+# install(DIRECTORY ${CEREAL_INC_DIR}/
+#     DESTINATION ${IPCL_INSTALL_INCLUDEDIR}
+#     FILES_MATCHING
+#     PATTERN "*.hpp"
+#     PATTERN "*.h"
+# )
 
 # IPP-Crypto (third party dep): include and install definition
-if(NOT ippcp_FOUND)
-  add_dependencies(ipcl ext_ipp-crypto)
-endif()
+# if(NOT ippcp_FOUND)
+#   add_dependencies(ipcl ext_ipp-crypto)
+# endif()
 
 target_include_directories(ipcl
     PUBLIC $<BUILD_INTERFACE:${IPPCRYPTO_INC_DIR}>
     PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/ipcl/ippcrypto>
 )
 
-install(DIRECTORY ${IPPCRYPTO_INC_DIR}/
-    DESTINATION ${IPCL_INSTALL_INCLUDEDIR}/ippcrypto
-    FILES_MATCHING
-    PATTERN "*.hpp"
-    PATTERN "*.h"
-)
+# install(DIRECTORY ${IPPCRYPTO_INC_DIR}/
+#     DESTINATION ${IPCL_INSTALL_INCLUDEDIR}/ippcrypto
+#     FILES_MATCHING
+#     PATTERN "*.hpp"
+#     PATTERN "*.h"
+# )
 
 # include and install definition of cpu_features
 if(IPCL_DETECT_CPU_RUNTIME)
-- 
2.39.3

